cmd/kubeadm/app/cmd/phases/init/kubeletready.go

```diff
@@ -0,0 +1,141 @@
+		Example: kubeletReadyPhaseExample,
+		Phases: []workflow.Phase{
+			{
+				Name:           "all",
+				Short:          "Run all kubelet-ready phases",
+				InheritFlags:   []string{options.CfgPath},
+				Example:        kubeletReadyPhaseExample,
+				RunAllSiblings: true,
+			},
+			{
+				Name:         "rotate-client-certs",
+				Short:        "Enable kubelet client certificate rotation",
+				InheritFlags: []string{options.CfgPath},
+				Run:          runKubeletReadyRotateCerts,
+			},
+		},
+	}
+}
+
+// runKubeletReadyRotateCerts detects if the kubelet certificate rotation is enabled
+// and updates the kubelet.conf file to point to a rotatable certificate and key for the
+// Node user.
+func runKubeletReadyRotateCerts(c workflow.RunData) error {
+	data, ok := c.(InitData)
+	if !ok {
+		return errors.New("kubelet-start phase invoked with an invalid data struct")
+	}
+
+	// Check if the user has added the kubelet --cert-dir flag.
+	// If yes, use that path, else use the kubeadm provided value.
+	cfg := data.Cfg()
+	pkiPath := filepath.Join(data.KubeletDir(), "pki")
+	val, ok := cfg.NodeRegistration.KubeletExtraArgs["cert-dir"]
+	if ok {
+		pkiPath = val
+	}
+
+	// Check for the existence of the kubelet-client-current.pem file in the kubelet certificate directory.
+	rotate := false
+	pemPath := filepath.Join(pkiPath, "kubelet-client-current.pem")
+	if _, err := os.Stat(pemPath); err == nil {
+		klog.V(1).Infof("[kubelet-ready] Assuming that kubelet client certificate rotation is enabled: found %q", pemPath)
+		rotate = true
+	} else {
+		klog.V(1).Infof("[kubelet-ready] Assuming that kubelet client certificate rotation is disabled: %v", err)
+	}
+
+	// Exit early if rotation is disabled.
+	if !rotate {
+		return nil
+	}
+
+	kubeconfigPath := filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.KubeletKubeConfigFileName)
+	fmt.Printf("[kubelet-ready] Updating %q to point to a rotatable kubelet client certificate and key\n", kubeconfigPath)
+
+	// Exit early if dry-running is enabled.
+	if data.DryRun() {
+		return nil
+	}
+
+	// Load the kubeconfig from disk.
+	kubeconfig, err := clientcmd.LoadFromFile(kubeconfigPath)
+	if err != nil {
+		return errors.Wrapf(err, "could not load %q", kubeconfigPath)
+	}
+
+	// Perform basic validation. The errors here can only happen if the kubelet.conf was corrupted.
+	userName := fmt.Sprintf("%s%s", kubeadmconstants.NodesUserPrefix, cfg.NodeRegistration.Name)
+	validationError := errors.Errorf("the file %q does not contain authentication for user %q", kubeconfigPath, cfg.NodeRegistration.Name)
+	if kubeconfig.AuthInfos == nil {
+		return validationError
+	}
+	info, ok := kubeconfig.AuthInfos[userName]
+	if !ok {
+		return validationError
+	}
+
+	// Update the client certificate and key of the node authorizer to point to the PEM symbolic link.
+	info.ClientKeyData = []byte{}
+	info.ClientCertificateData = []byte{}
+	info.ClientKey = pemPath
```

Comment:
Only for double-check. 
Is it correct to have both `ClientKey` and  `ClientCertificate` pointing to the same file?

Reply:
If its a pem that has both inside, its ok